osrf/vrx
# Wayfinding plugin

## Summary
Gazebosim port of the wayfinding plugin.

## Functionality
* Launches a world with a wamv in it.
* Gets waypoints from the sdf.
* Publishes them to `/vrx/wayfinding/waypoints'
* Gets the position of the wamv and calculates the minimum pose error achieved so far from the vehicle to each of the waypoints.
* Publishes the current minimums to `/vrx/wayfinding/min_errors`
* Publishes the mean of the minimums to `/vrx/wayfinding/mean_error` 

## To Test
Run each of the commands below in a separate terminal.
```
ros2 launch vrx_ros competition_local.launch.py ign_args:="-v 4 -r wayfinding_task.sdf"
```
to launch the wayfinding task. Then run:
```
ign topic --echo -t /vrx/task/info
```
to verify that the task info message is working correctly and reporting a score. This score should match the output of
```
ign topic --echo -t /vrx/wayfinding/mean_error
```
Next run:
```
ign topic --echo -t /vrx/wayfinding/waypoints
```
to see the waypoints. They should match the values listed in the `[wayfinding_task.sdf](https://github.com/osrf/vrx/blob/f732e8e6a8139a660db60edae513653990fa3274/vrx_ign/worlds/wayfinding_task.sdf)`:
```
  <waypoint>
    <pose>-33.7226766699 150.6740630167 1.21756121843</pose>
  </waypoint>
  
  <waypoint>
    <pose>-33.7220832884 150.67379127877 1.0</pose>
  </waypoint>
  
  <waypoint>
    <pose>-33.7226013209 150.67504609858 1.0</pose>
  </waypoint>
</waypoints>
```
Instead of yaw values, however, they will give orientation as a quaternion. I used this [online rotation converter](https://www.andre-gaschler.com/rotationconverter/) to check that a yaw of `1.21756121843` should correspond to the quaternion `[ 0, 0, 0.5718676, 0.820346 ]` and a yaw of `1` should correspond to `[ 0, 0, 0.4794255, 0.8775826 ]`. Finally, run  
```
ign topic --echo -t /vrx/wayfinding/min_errors
```
to see the individual min errors so far for each of the 3 waypoints. Now use the entity tree menu to select the wamv, then use the component inspector to move the wamv to the local coordinates corresponding to the waypoints:
```
X = -525.794 Y = 171.5 Yaw = 1.21756
X = -550.984 Y = 237.315 Yaw = 1
X = -434.679 Y = 179.863 Yaw = 1
```
Confirm that this causes the error to go to 0, or very close.

## Future Work
* Get the waypoint markers working. These should be implemented now but are not showing up for some reason. @j-herman notes a similar problem in PR #502, so I think we should save it for a separate PR.
